[quant][core][gpu][eager] Improved quantized conv operator in cudnn

dzdang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #72991
* __->__ #72770

Summary: This PR improves upon PR70622 by removing the call to_make_per_tensor_quantized_tensor
and directly creating a quantized int8 tensor that is passed into raw_cudnn_convolution_forward as
opposed to a non-quantized int8 tensor.

Differential Revision: [D34243926](https://our.internmc.facebook.com/intern/diff/D34243926)

dzdang committed:
test

[ghstack-poisoned]

dzdang committed:
Update on "[quant][core][gpu][eager] Improved quantized conv operator in cudnn"


Summary: This PR improves upon PR70622 by removing the call to_make_per_tensor_quantized_tensor
and directly creating a quantized int8 tensor that is passed into raw_cudnn_convolution_forward as
opposed to a non-quantized int8 tensor.

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34243926).

dzdang committed:
Update on "[quant][core][gpu][eager] Improved quantized conv operator in cudnn"


Summary: This PR improves upon PR70622 by removing the call to_make_per_tensor_quantized_tensor
and directly creating a quantized int8 tensor that is passed into raw_cudnn_convolution_forward as
opposed to a non-quantized int8 tensor.

Differential Revision: [D34243926](https://our.internmc.facebook.com/intern/diff/D34243926)

[ghstack-poisoned]

dzdang committed:
Update on "[quant][core][gpu][eager] Improved quantized conv operator in cudnn"


Summary: This PR improves upon PR70622 by removing the call to_make_per_tensor_quantized_tensor
and directly creating a quantized int8 tensor that is passed into raw_cudnn_convolution_forward as
opposed to a non-quantized int8 tensor.

Differential Revision: [D34243926](https://our.internmc.facebook.com/intern/diff/D34243926)

[ghstack-poisoned]

dzdang committed:
Update on "[quant][core][gpu][eager] Improved quantized conv operator in cudnn"


Summary: This PR improves upon PR70622 by removing the call to_make_per_tensor_quantized_tensor
and directly creating a quantized int8 tensor that is passed into raw_cudnn_convolution_forward as
opposed to a non-quantized int8 tensor.

Differential Revision: [D34243926](https://our.internmc.facebook.com/intern/diff/D34243926)

[ghstack-poisoned]

dzdang committed:
Update on "[quant][core][gpu][eager] Improved quantized conv operator in cudnn"


Summary: This PR improves upon PR70622 by removing the call to_make_per_tensor_quantized_tensor
and directly creating a quantized int8 tensor that is passed into raw_cudnn_convolution_forward as
opposed to a non-quantized int8 tensor.

Differential Revision: [D34243926](https://our.internmc.facebook.com/intern/diff/D34243926)

[ghstack-poisoned]

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/Conv.cpp:
```diff
@@ -201,7 +212,7 @@ void raw_cudnn_convolution_forward_out(
     auto workspace_size = 0;
     auto workspace = at::empty({workspace_size}, input.options().dtype(kByte));
-    void *data_ptrs[] = {reinterpret_cast<int8_t*>(input.data_ptr()), conv_output.data_ptr(), reinterpret_cast<int8_t*>(weight.data_ptr()), requantize_multiplier_tensor.data_ptr(), output.data_ptr()};
+    void *data_ptrs[] = {reinterpret_cast<int8_t*>(input.data_ptr()), conv_output.data_ptr(), reinterpret_cast<int8_t*>(weight.data_ptr()), requantize_multiplier_tensor.data_ptr(), quantized_output.data_ptr()};
```
we can do a reinterpret_cast for quantized_output.data_ptr() here as well I think:
```
reinterpret_cast<int8_t*>(quantized_output.data_ptr())
```
original data type should be `qint8_t*` but we can cast it to `int8_t*`

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/Conv.cpp:
```diff
@@ -232,7 +243,9 @@ void raw_cudnn_convolution_forward_out(
     .setbDesc(getTensorDescriptor(requantize_multiplier_tensor, 's', getAlignment(requantize_multiplier_tensor)))
-    .setyDesc(getTensorDescriptor(output, 'r', getAlignment(output)))
+    .setyDesc(getTensorDescriptor(quantized_output.sizes(), quantized_output.strides(), CUDNN_DATA_INT8, 'r', getAlignment(quantized_output)))
+    // I think this is inefficient. it makes a copy I believe. The above one can be used instead
+    // .setyDesc(getTensorDescriptor(quantized_output.int_repr(), 'r', getAlignment(quantized_output)))
```
we can remove this

jerryzh168 reviewed (approved):
looks great, had a few inline comments, please attach the new benchmark results as well

dzdang committed:
Update on "[quant][core][gpu][eager] Improved quantized conv operator in cudnn"


Summary: This PR improves upon PR70622 by removing the call to_make_per_tensor_quantized_tensor
and directly creating a quantized int8 tensor that is passed into raw_cudnn_convolution_forward as
opposed to a non-quantized int8 tensor.

Differential Revision: [D34243926](https://our.internmc.facebook.com/intern/diff/D34243926)

[ghstack-poisoned]

dzdang commented:
Current benchmark results (compare with https://github.com/pytorch/pytorch/pull/70622)
```
parameters: batch_size: 16 in_channel: 64 out_channel: 64 kernel_size: 3 height: 256 widht: 256
fp32 benchmark result:
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg     Self CUDA   Self CUDA %    CUDA total  CUDA time avg    # of Calls  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                  cudaDeviceSynchronize        87.38%      64.031ms        87.38%      64.031ms      64.031ms       0.000us         0.00%       0.000us       0.000us             1  
                                aten::cudnn_convolution        11.07%       8.113ms        11.72%       8.588ms     429.400us      59.644ms        83.33%      59.644ms       2.982ms            20  
                                          ProfilerStep*         0.47%     344.000us        12.48%       9.142ms     457.100us       0.000us         0.00%      59.644ms       2.982ms            20  
                                       cudaLaunchKernel         0.45%     329.000us         0.45%     329.000us       8.225us       0.000us         0.00%       0.000us       0.000us            40  
                                            aten::empty         0.24%     174.000us         0.24%     174.000us       2.900us       0.000us         0.00%       0.000us       0.000us            60  
                                     aten::_convolution         0.11%      81.000us        11.83%       8.669ms     433.450us       0.000us         0.00%      59.644ms       2.982ms            20  
                                      aten::convolution         0.10%      72.000us        11.93%       8.741ms     437.050us       0.000us         0.00%      59.644ms       2.982ms            20  
                                            aten::zeros         0.09%      65.000us         0.14%     105.000us       5.250us       0.000us         0.00%       0.000us       0.000us            20  
                                           aten::conv2d         0.05%      40.000us        11.98%       8.781ms     439.050us       0.000us         0.00%      59.644ms       2.982ms            20  
                                        cudaEventRecord         0.03%      23.000us         0.03%      23.000us       1.150us       0.000us         0.00%       0.000us       0.000us            20  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 73.278ms
Self CUDA time total: 71.579ms

fp16 benchmark result:
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg     Self CUDA   Self CUDA %    CUDA total  CUDA time avg    # of Calls  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                  cudaDeviceSynchronize        99.43%        1.346s        99.43%        1.346s        1.346s       0.000us         0.00%       0.000us       0.000us             1  
                                aten::cudnn_convolution         0.47%       6.298ms         0.52%       7.001ms     350.050us     898.399ms        80.00%     898.399ms      44.920ms            20  
                                       cudaLaunchKernel         0.04%     508.000us         0.04%     508.000us       8.467us       0.000us         0.00%       0.000us       0.000us            60  
                                          ProfilerStep*         0.03%     375.000us         0.56%       7.597ms     379.850us       0.000us         0.00%     898.399ms      44.920ms            20  
                                            aten::empty         0.02%     206.000us         0.02%     206.000us       3.433us       0.000us         0.00%       0.000us       0.000us            60  
                                     aten::_convolution         0.01%      91.000us         0.52%       7.092ms     354.600us       0.000us         0.00%     898.399ms      44.920ms            20  
                                      aten::convolution         0.01%      70.000us         0.53%       7.162ms     358.100us       0.000us         0.00%     898.399ms      44.920ms            20  
                                            aten::zeros         0.00%      58.000us         0.01%     115.000us       5.750us       0.000us         0.00%       0.000us       0.000us            20  
                                           aten::conv2d         0.00%      43.000us         0.53%       7.205ms     360.250us       0.000us         0.00%     898.399ms      44.920ms            20  
                                   cudaFuncSetAttribute         0.00%      30.000us         0.00%      30.000us       0.750us       0.000us         0.00%       0.000us       0.000us            40  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 1.354s
Self CUDA time total: 1.123s

int8 benchmark result:
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg     Self CUDA   Self CUDA %    CUDA total  CUDA time avg    # of Calls  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                  cudaDeviceSynchronize        85.39%      19.085ms        85.39%      19.085ms      19.085ms       0.000us         0.00%       0.000us       0.000us             1  
                                quantized::conv2d_cudnn         3.88%     867.000us        10.90%       2.436ms     121.800us       0.000us         0.00%       6.111ms     305.550us            20  
                                          ProfilerStep*         3.22%     719.000us        14.20%       3.174ms     158.700us       0.000us         0.00%       6.111ms     305.550us            20  
                                            aten::empty         2.27%     508.000us         2.27%     508.000us       3.629us       0.000us         0.00%       0.000us       0.000us           140  
                                       cudaLaunchKernel         1.99%     445.000us         1.99%     445.000us       7.417us       0.000us         0.00%       0.000us       0.000us            60  
                                         aten::int_repr         1.48%     330.000us         3.85%     860.000us      21.500us       2.675ms        12.79%       2.675ms      66.875us            40  
                                            aten::fill_         0.56%     126.000us         1.11%     249.000us      12.450us       3.436ms        16.43%       3.436ms     171.800us            20  
                          aten::_empty_affine_quantized         0.47%     105.000us         0.47%     105.000us       5.250us       0.000us         0.00%       0.000us       0.000us            20  
                                          aten::q_scale         0.38%      85.000us         0.38%      85.000us       2.125us       0.000us         0.00%       0.000us       0.000us            40  
                                            aten::zeros         0.25%      55.000us         0.41%      91.000us       4.550us       0.000us         0.00%       0.000us       0.000us            20  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 22.350ms
Self CUDA time total: 20.914ms
```

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34243926).
